src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs

```diff
@@ -0,0 +1,195 @@
+        //
+        // In the event that these two source either don't have filenames set or don't follow a coherent hierarchy,
+        // we will just use the namespace verbatim.
+        internal static bool TryComputeNamespace(string source, DirectiveIRNode directive, out string @namespace)
+        {
+            var directiveSource = NormalizeDirectory(directive.Source?.FilePath);
+
+            var baseNamespace = directive.Tokens.First().Content;
+            if (string.IsNullOrEmpty(source) || directiveSource == null)
+            {
+                // No sources, can't compute a suffix.
+                @namespace = baseNamespace;
+                return false;
+            }
+
+            if (!source.StartsWith(directiveSource) || source.Length <= directiveSource.Length)
+            {
+                // The imports are not from the directory hierarchy, can't compute a suffix.
+                @namespace = baseNamespace;
+                return false;
+            }
+
+            // OK so that this point we know that the 'imports' file containing this directive is in the directory
+            // hierarchy of this soure file. This is the case where we can append a suffix to the baseNamespace.
+            //
+            // Everything so far has just been defensiveness on our part.
+
+            var builder = new StringBuilder(baseNamespace);
+
+            var segments = source.Substring(directiveSource.Length).Split('/', '\\');
+
+            // Skip the last segment because it's the filename.
+            for (var i = 0; i < segments.Length - 1; i++)
+            {
+                builder.Append('.');
+                builder.Append(segments[i]);
+            }
+
+            @namespace = builder.ToString();
+            return true;
+        }
+
+        // We want to normalize the path of the file containing the '@namespace' directive to just the containing
+        // directory with a trailing separator.
+        //
+        // Not using Path.GetDirectoryName here because it doesn't meet these requirements, and we want to handle
+        // both 'view engine' style paths and absolute paths.
+        //
+        // We also don't normalize the separators here. We expect that all documents are using a consistent style of path.
+        // 
+        // If we can't normalize the path, we just return null so it will be ignored.
+        private static string NormalizeDirectory(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                return null;
+            }
+
+            var lastSeparator = path.LastIndexOfAny(Separators);
+            if (lastSeparator == -1)
+            {
+                return null;
+            }
+
+            // Includes the separator
+            return path.Substring(0, lastSeparator + 1);
+        }
+
+        private class NamespaceDirectiveNodeVisitor : RazorIRNodeWalker
```

Comment:
Any reason to not have 1 visitor that extracts the directives, namespaces and classes in a single pass?